pkg/client/supplier/client.go

```diff
@@ -63,34 +66,35 @@ func NewSupplierClient(
 // the transaction is included in a block or times out.
 func (sClient *supplierClient) SubmitProofs(
 	ctx context.Context,
-	sessionProofs []*relayer.SessionProof,
+	proofMsgs ...client.MsgSubmitProof,
 ) error {
+	sClient.pendingTxMu.Lock()
+	defer sClient.pendingTxMu.Unlock()
 	logger := polylog.Ctx(ctx)
 
-	msgs := make([]cosmostypes.Msg, len(sessionProofs))
-
-	for i, sessionProof := range sessionProofs {
-		// TODO(@bryanchriswhite): reconcile splitting of supplier & proof modules
-		//  with off-chain pkgs/nomenclature.
-		msgs[i] = &prooftypes.MsgSubmitProof{
-			SupplierAddress: sessionProof.SupplierAddress.String(),
-			SessionHeader:   sessionProof.SessionHeader,
-			Proof:           sessionProof.ProofBz,
-		}
+	msgs := make([]cosmostypes.Msg, 0, len(proofMsgs))
+	for _, p := range proofMsgs {
+		msgs = append(msgs, p)
 	}
 
+	// TODO(@bryanchriswhite): reconcile splitting of supplier & proof modules
+	//  with off-chain pkgs/nomenclature.
 	eitherErr := sClient.txClient.SignAndBroadcast(ctx, msgs...)
 	err, errCh := eitherErr.SyncOrAsyncError()
 	if err != nil {
 		return err
 	}
 
-	for _, sessionProof := range sessionProofs {
-		sessionHeader := sessionProof.SessionHeader
+	for _, p := range proofMsgs {
+		// Type casting does not need to be checked here since the concrete type is
+		// guaranteed to implement the interface which is just an identity for the
+		// concrete type.
+		proofMsg, _ := p.(*prooftypes.MsgSubmitProof)
```

Comment:
**TL;DR, I'm not a fan of the `I` pre/postfix but would be willing to concede on this detail as it's part of a temporary solution. I think it would be sufficient to move these interfaces to an `expected_interfaces.go` and add them on an as-needed basis. A more comprehensive refactoring will be necessary to fully resolve this.**

I think that, ideally, the `pkg` package SHOULD NOT depend on anything from `x/...`. The current issue is that the generated go protobuf types are used both off- and on-chain. So long as go protobuf types are being generated into `x/...`, `pkg/client/...` (and possibly other packages) will either have dependency cycles which they will have to work around, or such dependency cycles will be looming as several interface types in `pkg/client/interface.go` already reference types from `x/<module>/types`.

In my view, the correct solution here should involve moving the generated go protobuf types to a package which is not a branch of either `pkg/` or `x/` (e.g., `proto/types/<module>`). This would:

1. Resolve this class dependency cycles; eliminating the need for redundant interface message types (i.e., client <-> proto).
1. Eliminate on-chain pkg dependencies from leaking through client pkg. This has implications for potential binary size optimzations for off-chain binaries which depend on our protobuf types (e.g., Shannon SDK consumers).